connectivity/check/action.go

```diff
@@ -721,18 +721,31 @@ func (a *Action) GetIngressRequirements(p FlowParameters) []filters.FlowSetRequi
 }
 
 // waitForRelay polls the server status from Relay until either it's connected to all the Hubble
-// instances (success) or the context is cancelled (failure).
+// instances, and it started receiving flows (success) or the context is cancelled (failure).
 func (a *Action) waitForRelay(ctx context.Context, client observer.ObserverClient) error {
 	for {
-		res, err := client.ServerStatus(ctx, &observer.ServerStatusRequest{})
-		if err == nil && (res.NumUnavailableNodes == nil || res.NumUnavailableNodes.Value == 0) {
-			// This means all the nodes are available.
-			return nil
+		// Sometimes ServerStatus request just gets stuck until the context gets cancelled. Call
+		// ServerStatus with a shorter timeout than ctx so that we get to try it multiple times.
+		timeoutContext, cancel := context.WithTimeout(ctx, 3*time.Second)
+		res, err := client.ServerStatus(timeoutContext, &observer.ServerStatusRequest{})
+		cancel()
```

Comment:
NIT:
This method looks a bit tricky, so I just wonder what if we move the status check into a new private method:
```
func (a *Action) isServerStatusOK(ctx context.Context, client observer.ObserverClient) bool {
	// Sometimes ServerStatus request just gets stuck until the context gets cancelled. Call
	// ServerStatus with a shorter timeout than ctx so that we get to try it multiple times.
	timeoutContext, cancel := context.WithTimeout(ctx, 3*time.Second)
	defer cancel()
	res, err := client.ServerStatus(timeoutContext, &observer.ServerStatusRequest{})
	if err != nil {
		a.Debugf("hubble relay server status failed: %v", err)
		return false
	}
	a.Debugf("hubble relay server status %+v", res)
	// This means all the nodes are available, and Hubble Relay started receiving flows.
	// Ideally we should check res.NumConnectedNodes matches the expected number of Cilium
	// nodes instead of relying on res.NumUnavailableNodes, but I don't know if that information
	// is available to us.
	return (res.NumUnavailableNodes == nil || res.NumUnavailableNodes.Value == 0) && res.NumFlows > 0
}
```

and then the current method will look like:
```
func (a *Action) waitForRelay(ctx context.Context, client observer.ObserverClient) error {
	for {
		if a.isServerStatusOK(ctx, client) {
			return nil
		}
		select {
		case <-ctx.Done():
			return fmt.Errorf("hubble server status failure: %w", ctx.Err())
		case <-time.After(time.Second):
			a.Debugf("retrying hubble relay server status request")
		}
	}
}
```

Just a suggestion, the current implementation looks good.

Reply:
updated 👍